tools/clean.sh

```diff
@@ -0,0 +1,109 @@
+
+This command processes spaces and tabs in each line of a file
+according to the following subcommands:
+  D - Delete all spaces and tabs
+  B - Delete spaces and tabs at the beginning of each line
+  E - Delete spaces and tabs at the end of each line
+  R - Reduce multiple spaces and tabs to one space
+  Q - Do not alter characters in quotes (ignored by T and U)
+  T - (Tab) Convert spaces to equivalent tabs (unsupported)
+  U - (Untab) Convert tabs to equivalent spaces (unsupported)
+Some other subcommands are also available:
+  P - Trim parity (8th) bit from each character
+  G - Discard garbage characters CR,FF,ESC,BS
+  C - Convert to upper case
+  L - Convert to lower case
+  V - Convert VT220 screen print frame graphics to -,|,+ characters
+Subcommands are separated by spaces, e.g., B E R Q
+Syntax:  clean.sh subcmd subcmd ... < INFILE > OUTFILE
+HELP
+}
+
+if [ $# -eq 1 ] && [ "$1" = "-h" ]; then usage; exit; fi
+if [ $# -eq 0 ]; then usage; exit 1; fi
+
+for var in "$@"; do
+  case $var in
+    D | d) allDiscard=1;;
+    B | b) leadDiscard=1;;
+    E | e) trailDiscard=1;;
+    R | r) reduce=1;;
+    Q | q) quote=1;;
+    T | t) echo "T (tab) flag unsupported"; exit 2;;
+    U | u) echo "U (untab) flag unsupported"; exit 2;;
+    P | p) trim=1;;
+    G | g) discardCtrl=1;;
+    C | c) caseFlag=1;;
+    L | l) lowercase=1;;
+    V | v) screen=1;;
+  esac
+done
+
+awk \
+  -v trim=$trim \
+  -v allDiscard=$allDiscard \
+  -v discardCtrl=$discardCtrl \
+  -v leadDiscard=$leadDiscard \
+  -v reduce=$reduce \
+  -v caseFlag=$caseFlag \
+  -v trailDiscard=$trailDiscard \
+  -v quote=$quote \
+  -v lowercase=$lowercase \
+  -v tab=$tab \
+  -v untab=$untab \
+  -v screen=$screen \
+'BEGIN {
+  FS="";
+  if (trim || caseFlag || lowercase) {
+    for (n=0; n<256; n++) {
+      ch = sprintf("%c", n);
+      chr[n] = ch;
+      ord[ch] = n;
+    }
+  }
+  if (discardCtrl) {
+    discardArr["\r"] = 1;
+    discardArr["\n"] = 1;
+    discardArr["\f"] = 1;
+    discardArr["\x27"] = 1;
+    discardArr["\x08"] = 1;
+  }
+  if (screen) {
+    gfxArr["\xEA"] = gfxArr["\xEB"] = gfxArr["\xEC"] = gfxArr["\xED"] = "+";
+    gfxArr["\xDA"] = gfxArr["\xD9"] = gfxArr["\xBF"] = gfxArr["\xC0"] = "+";
+    gfxArr["\xF1"] = gfxArr["\xC4"] = "-";
+    gfxArr["\xF8"] = gfxArr["\xA6"] = gfxArr["\xB3"] = "|";
+  }
+} {
+  if (leadDiscard)
+    gsub(/^[ \t]+/, "");
+  inQuote = 0;
+  out = "";
+  for (i=1;i<=NF;i++) {
+    if (quote && ($i == "\"" || $i == "\x39"))
+      inQuote = !inQuote;
+    if (inQuote) {
+      out=out $i;
+      continue;
+    }
+    if (allDiscard && ($i == " " || $i == "\t")) continue;
+    if (trim) $i = chr[and(ord[$i], 0x7f)];
+    if (discardCtrl && discardArr[$i]) continue;
+    if (caseFlag && ord[$i] < 128) $i = toupper($i);
+    if (lowercase && ord[$i] < 128) $i = tolower($i);
+    if (screen && gfxArr[$i]) $i = gfxArr[$i];
+    out=out $i;
+  }
+  if (trailDiscard)
```

Comment:
I wonder whether one can have this at the beginning: it eliminates checks in the for loop

Reply:
That kind of motion is not likely to preserve behavior, since the for loop can do quite a few things to the input

Reply:
OK, but then move the check for leading spaces here, too.